swim/self_evict_test.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package swim
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/mock"
+
+	"github.com/stretchr/testify/suite"
+)
+
+type SelfEvictTestSuite struct {
+	suite.Suite
+	tnode, tpeer *testNode
+	node, peer   *Node
+}
+
+func (s *SelfEvictTestSuite) SetupTest() {
+	s.tnode = newChannelNode(s.T())
+	s.node = s.tnode.node
+	s.tpeer = newChannelNode(s.T())
+	s.peer = s.tpeer.node
+
+	bootstrapNodes(s.T(), s.tnode, s.tpeer)
+	waitForConvergence(s.T(), 5*time.Second, s.tnode, s.tpeer)
+}
+
+func (s *SelfEvictTestSuite) TearDownTest() {
+	destroyNodes(s.tnode, s.tpeer)
+}
+
+func (s *SelfEvictTestSuite) TestSelfEvict_RegisterSelfEvictHook() {
+	// have some hooks
+	hook1 := new(MockSelfEvictHook)
+	hook1.On("Name").Return("hook1")
+
+	hookDuplicate := new(MockSelfEvictHook)
+	hookDuplicate.On("Name").Return("hook1")
+
+	hook2 := new(MockSelfEvictHook)
+	hook2.On("Name").Return("hook2")
+
+	err := s.node.RegisterSelfEvictHook(hook1)
+	s.Assert().NoError(err, "expected no error when a hook gets attached for the first time")
+
+	err = s.node.RegisterSelfEvictHook(hook2)
+	s.Assert().NoError(err, "expected no error when a second hook gets attached for the first time")
+
+	err = s.node.RegisterSelfEvictHook(hookDuplicate)
+	s.Assert().Error(err, "expected an error when a duplicate hook gets attached to ringpop")
+}
+
+func (s *SelfEvictTestSuite) TestSelfEvict_SelfEvict() {
+	hooks := &MockSelfEvictHook{}
+	hooks.On("Name").Return("hooks")
+
+	hooks.On("PreEvict").Run(func(args mock.Arguments) {
+		phase := s.node.selfEvict.currentPhase()
```

Comment:
I think we want coverage of the function anyway. I don't know if we need separate tests for it. I'm definitely in favor of _NOT_ exporting the function.